### PR TITLE
[MRG+1] ENH:  morph between hemispheres

### PR DESCRIPTION
**Pull request by christianbrodbeck**
Addresses #4360 

@Eric89GXL what do you think of the API? An example is below.

The `'lh'`/`'rh'` tags allow specifying which direction to map (I don't think mirroring both hemispheres will be useful). An alternative would be that the user should specify vertices for only the required hemispheres, like `[[], rh_vertices], [lh_vertices, []]` to morph from right to left; in that case we could actually remove the reg_from and reg_to parameters if all we want to add is xhemi morphing, but this solution is less explicit than the parameters.

Concerning testing, actually testing the x-hemi registration would require adding a new registration file to the dataset. Alternatively I could just add a test loading a morph matrix for only one hemisphere.

screen shot 2017-07-14 at 12 36 15 pm

```
# Data from examples/MNE-sample-data/MEG/sample
import mne

sdir = '/Applications/freesurfer/subjects'

stc = mne.read_source_estimate('fsaverage_audvis-meg-eeg-lh.stc', 'fsaverage')
stc_sym = stc.morph('fsaverage_sym', 3, subjects_dir=sdir)

mm = mne.compute_morph_matrix('fsaverage_sym', 'fsaverage_sym', stc.vertices, stc.vertices, subjects_dir=sdir, reg_from='rh.sphere.left_right', reg_to='lh.sphere.left_right')
stc_rh_on_lh = mne.SourceEstimate(mm * stc_sym.rh_data, [stc.vertices[0], np.empty(0)], stc.tmin, stc.tstep, 'fsaverage_sym')

# plot
size = (300, 300)
brain_rh = stc_sym.plot(hemi='rh', subjects_dir=sdir, initial_time=0.09, size=size)
brain_lh = stc_rh_on_lh.plot(hemi='lh', subjects_dir=sdir, initial_time=0.09, size=size)

```

### CODE REVIEW

**Comment by Eric89GXL** [comment]
> An alternative would be that the user should specify vertices for only the required hemispheres, 

Seems a bit too magical to me.

> Concerning testing, actually testing the x-hemi registration would require adding a new registration file to the dataset. 

The ~6MB is probably worth adding to the `mne-testing-data` repo. You can for example have activity in the center vertex of some left hemisphere label (STS?) and ensure it lands in the right hemi label.

> The 'lh'/'rh' tags allow specifying which direction to map (I don't think mirroring both hemispheres will be useful)

I don't see the harm in mirroring both, then you can think of it as a simple hemi-swap operation (which is what the `left_right` in the filename suggests anyway).

> mm = mne.compute_morph_matrix('fsaverage_sym', 'fsaverage_sym', stc.vertices, stc.vertices, subjects_dir=sdir, reg_from='rh.sphere.left_right', reg_to='lh.sphere.left_right')

This seems a bit difficult to read because it has `left_right` and a `lh` or `rh` in the name. What if we only allowed entries `reg='sphere'` or `reg='sphere.left_right'` for now, where the former is what we're used to and the latter swaps the hemispheres (doing something different with lh/rh as necessary)?

**Comment by agramfort** [comment]
I am sorry @christianbrodbeck I don't understand what you try to do? can you dumb it down for me?

**Comment by Eric89GXL** [comment]
Did you read #4360? It's in there. Basically `FreeSurfer` provides a way to morph activity from one hemi to the other for a special subject `fsaverage_sym`.

**Comment by agramfort** [comment]
ok so it's just exposing freesurfer capabilities? we don't reinvent the
wheel?

**Comment by christianbrodbeck** [comment]
@agramfort yes, basically FreeSurfer can create morph-maps that we can't yet use

**Comment by agramfort** [comment]
ok no objection then

**Comment by christianbrodbeck** [comment]
> The ~6MB is probably worth adding to the mne-testing-data repo

So I should make a separate PR into that repo?

>I don't see the harm in mirroring both, then you can think of it as a simple hemi-swap operation (which is what the `left_right` in the filename suggests anyway).

I don't think there is harm in it :), but the main use case I see in mirroring one hemisphere to the other so that the hemispheres can be compared for a test of lateralization. Do you have a use case in mind for mirroring both at once? 

> This seems a bit difficult to read because it has `left_right` and a `lh` or `rh` in the name. What if we only allowed entries `reg='sphere'` or `reg='sphere.left_right'` for now, where the former is what we're used to and the latter swaps the hemispheres (doing something different with lh/rh as necessary)?

This would work too. Without the 'lh'/'rh' we would not know whether the user wants to mirror left to right or vice versa from the `reg` parameter, so in order to get a morph matrix for one hemisphere one would set vertices appropriately – the new line would then be:

`mm = mne.compute_morph_matrix('fsaverage_sym', 'fsaverage_sym', [[], stc.vertices[1]], [stc.vertices[0], []], subjects_dir=sdir, reg='sphere.left_right')`

I agree it looks simpler. If we only have two options we could also just use `xhemi=True`?.

We would be covering the case of [mirroring within subject](http://www.mail-archive.com/freesurfer@nmr.mgh.harvard.edu/msg25832.html). FreeSurfer can also [register subjects directly to fsaverage_sym](https://surfer.nmr.mgh.harvard.edu/fswiki/Xhemi#Maptherighthemisphereofanindividualtotheleftofsameperson) which someone might want in the future, and using `xhemi=True` would actually make adding that option more straight forward because it does not use `sphere.left_right` but could be recognized through `xhemi=True` and `subject_from != subject_to`.

**Comment by christianbrodbeck** [comment]
New test script:

```
# Data from examples/MNE-sample-data/MEG/sample
import mne
import numpy as np

sdir = '/Applications/freesurfer/subjects'

stc = mne.read_source_estimate('fsaverage_audvis-meg-eeg-lh.stc', 'fsaverage')
stc_sym = stc.morph('fsaverage_sym', 3, subjects_dir=sdir)

mm = mne.compute_morph_matrix('fsaverage_sym', 'fsaverage_sym', [[], stc.vertices[1]], [stc.vertices[0], []], subjects_dir=sdir, xhemi=True)
stc_rh_on_lh = mne.SourceEstimate(mm * stc_sym.rh_data, [stc.vertices[0], np.empty(0)], stc.tmin, stc.tstep, 'fsaverage_sym')

# plot
size = (300, 300)
brain_rh = stc_sym.plot(hemi='rh', subjects_dir=sdir, initial_time=0.09, size=size)
brain_lh = stc_rh_on_lh.plot(hemi='lh', subjects_dir=sdir, initial_time=0.09, size=size)


# opposite direction
mm = mne.compute_morph_matrix('fsaverage_sym', 'fsaverage_sym', [stc.vertices[0], []], [[], stc.vertices[1]], subjects_dir=sdir, xhemi=True)
stc_lh_on_rh = mne.SourceEstimate(mm * stc_sym.lh_data, [np.empty(0), stc.vertices[1]], stc.tmin, stc.tstep, 'fsaverage_sym')

# plot
brain_lh = stc_sym.plot(hemi='lh', subjects_dir=sdir, initial_time=0.09, size=size)
brain_rh = stc_lh_on_rh.plot(hemi='rh', subjects_dir=sdir, initial_time=0.09, size=size)
```

**Comment by christianbrodbeck** [comment]
Fsaverage in FreeSurfer 6 contains left_right maps, since the testing data already contains other components of fsaverage I assume it's ok to add those files?

**Comment by agramfort** [comment]
yes I would say so

**Comment by Eric89GXL** [comment]
Yes it's okay to add them, make a PR to `mne-testing-data` and maybe also `mne-sample-data`. This will change the example, then, right?

**Comment by christianbrodbeck** [comment]
> This will change the example, then, right?

Yes, you would not need to morph to `fsaverage_sym` first (but this step is recommended for statistical inter hemisphere comparison [here](http://www.mail-archive.com/freesurfer@nmr.mgh.harvard.edu/msg25832.html))

Should I document this in the manual, or with an example, or just a note in `compute_morph_matrix`?

**Comment by Eric89GXL** [comment]
I would say:

1. Add to the `Notes` section of `compute_morph_matrix` how to use the registration parameter(s), and `.. note::` that `fsaverage_sym` should be used for statistical purposes.
2. Add a pointer to the `Notes` section in the docstring of those parameters
3. Add an example of how to use this functionality. Also have a `.. note::` about `fsaverage_sym` / statistics there, but you can point to the `compute_morph_matrix` docstring for more details to avoid too much copy-paste.

WDYT?

**Comment by christianbrodbeck** [comment]
Sounds good

How can the files be added to the `mne-sample-data`?

**Comment by Eric89GXL** [comment]
They are hosted on @dengemann's Amazon account -- @dengemann can you help?

**Comment by christianbrodbeck** [comment]
I've updated the doc-string of `compute_morph_matrix`. Maybe an example can be added later once the sample dataset can be updated?

**Comment by larsoner** [comment]
You can make an example using a combination of files from `sample` and `testing`, right?

**Comment by christianbrodbeck** [comment]
Good idea, I added an example. In order to make STC initialization less awkward I added https://github.com/mne-tools/mne-python/pull/4383/commits/ab93138d7a8b8ff737cc839b37f3a234af8a53f2, let me know id you think that's not a good idea; I was also wondering whether I should add `dtype=int`.

**Comment by christianbrodbeck** [comment]
@larsoner what do I need to do in addition to adding the `plot_....py` file to the examples subdirectory?

**Comment by larsoner** [comment]
If it's an example, nothing. If you are asking why `CircleCI` is failing, it could be the testing data is old, so I restarted without using cache. If that still fails, then we'll have to look closer at why it can't find the file.

**Comment by christianbrodbeck** [comment]
Hm, failed again with the same message. Locally I can `run` the example from iPython without error. How can I figure out what's happening? I haven't used sphinx gallery before.

**Comment by christianbrodbeck** [comment]
Does it need to be added under `doc/documentation.rst`? Looks like the other examples are listed there.

File: mne/tests/test_surface.py [hunk 1]
```diff
@@ -95,20 +95,27 @@ def test_make_morph_maps():
     for subject in ('sample', 'sample_ds', 'fsaverage_ds'):
         os.mkdir(op.join(tempdir, subject))
         os.mkdir(op.join(tempdir, subject, 'surf'))
+        regs = ('reg', 'left_right') if subject == 'fsaverage_ds' else ('reg',)
```
Comment (larsoner) [inline mne/tests/test_surface.py]: what is this conditional for `fsaverage_ds`?

Comment (larsoner) [reply mne/tests/test_surface.py]: (`fsaverage_ds` shouldn't be treated any differently within MNE from any other subject -- it just so happens that we've downsampled it for faster processing)

Comment (christianbrodbeck) [reply mne/tests/test_surface.py]: This is just copying files needed for a test so that the test does not write into the testing data folder.

**Comment by christianbrodbeck** [comment]
Aah, I only added the left_right registrations to `fsaverage_ds` to keep the testing data size lower, assuming we'd add the full resolution ones to the sample dataset. Should I add the full resolution ones to the testing data then?

Also how did you find the underlying error?

**Comment by larsoner** [comment]
Ahh sorry didn't notice that conditional was in a test, that should be fine

> Should I add the full resolution ones to the testing data then?

They probably fit better with the sample dataset. @dengemann can you help?

> Also how did you find the underlying error?

It's in the circleci log (whichever collapsed section precedes the regex
that looks for satisfactory completion)

**Comment by christianbrodbeck** [comment]
Ah, thanks!

**Comment by christianbrodbeck** [comment]
Should we defer the example until the dataset can be updated so that the changes can be merged before we get merge conflicts?

**Comment by agramfort** [comment]
fine with me

**Comment by larsoner** [comment]
We can update datasets much more easily now. I can roll the changes into my
currently open datasets PR. Okay with you to wait a week to get that in?

On Sep 24, 2017 10:15, "Alexandre Gramfort" <notifications@github.com>
wrote:

fine with me

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/mne-tools/mne-python/pull/4383#issuecomment-331712922>,
or mute the thread

**Comment by christianbrodbeck** [comment]
Sure, should we add a 0.15 milestone since after the dataset PR this should be ready?

File: examples/visualization/plot_xhemi.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import mne
+
+
+subjects_dir = mne.datasets.testing.data_path() + '/subjects'
+data_dir = mne.datasets.sample.data_path()
+stc_path = data_dir + '/MEG/sample/sample_audvis-meg-eeg'
+
+stc = mne.read_source_estimate(stc_path, 'sample')
+
+# First, morph the data to fsaverage because we have left-righ registrations
```
Comment (larsoner) [inline examples/visualization/plot_xhemi.py]: right

File: examples/visualization/plot_xhemi.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+# for fsaverage:
+stc = stc.morph('fsaverage', subjects_dir=subjects_dir)
```
Comment (larsoner) [inline examples/visualization/plot_xhemi.py]: this takes a while, just do `smooth=5`

**Comment by larsoner** [comment]
The necessary `MNE-sample-data` changes have been made in #4585. Could you try that PR to update your sample data, and see if you can refactor this PR to use the new `sample` files? You shouldn't need to use the `testing` dataset in the example anymore, and can add `fsavearge_sym` stuff, too, if you want.

**Comment by larsoner** [comment]
@christianbrodbeck any time for this one this week? If so we can get it into 0.15, if not we'll probably push to 0.16

**Comment by christianbrodbeck** [comment]
Yes I'll have time for it tomorrow afternoon

**Comment by christianbrodbeck** [comment]
On CircleCI I get `IOError: [Errno 2] No such file or directory: u'/home/ubuntu/mne_data/MNE-sample-data/subjects/fsaverage_sym/surf/lh.sphere.reg'`, I did not get that error locally, is it possible circle is using a cached sample dataset? Locally after `force_update`ing the sample data I have version 0.11, is that correct?

**Comment by larsoner** [comment]
Yes it probably has it cached. In the controls for the CircleCI build in the upper right you can select "Restart without cache"

**Comment by christianbrodbeck** [comment]
Rendered example: https://5958-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.7Vp0i4C/html/auto_examples/visualization/plot_xhemi.html#sphx-glr-auto-examples-visualization-plot-xhemi-py

**Comment by christianbrodbeck** [comment]
(It was the cache, thanks!)

**Comment by christianbrodbeck** [comment]
@larsoner tests passed and your comments are addressed, the link to the example is in the post just above

**Comment by agramfort** [comment]
thx @christianbrodbeck
